postgres/tests/test_deadlock.py

```diff
@@ -0,0 +1,119 @@
+            state = conn.poll()
+            if state == psycopg2.extensions.POLL_OK:
+                break
+            elif state == psycopg2.extensions.POLL_WRITE:
+                select.select([], [conn.fileno()], [])
+            elif state == psycopg2.extensions.POLL_READ:
+                select.select([conn.fileno()], [], [])
+            else:
+                raise psycopg2.OperationalError("poll() returned %s" % state)
+            time.sleep(0.1)
+
+    conn_args = { 'host': HOST, 'dbname': DB_NAME, 'user': "bob", 'password': "bob"}
+    conn1 = psycopg2.connect(**conn_args)
+    conn1.autocommit = False
+
+    conn2 = psycopg2.connect(**conn_args, async_=1)
+    wait(conn2)
+
+    appname = 'deadlock sess'
+    appname1 = appname + '1'
+    appname2 = appname + '2'
+    appname_sql = "SET application_name=%s"
+    update_sql =  "update personsdup1 set address = 'changed' where personid = %s"
+
+    deadlock_count_sql = "select deadlocks from pg_stat_database where datname = %s"
+    cursor.execute(deadlock_count_sql, (DB_NAME, ))
+    deadlocks_before = cursor.fetchone()[0]
+
+    cur1 = conn1.cursor()
+    cur1.execute(appname_sql, (appname1,))
+    cur1.execute(update_sql, (1,) )
+
+    cur2 = conn2.cursor()
+    cur2.execute("""SET application_name=%s;
+begin transaction;
+{};
+{};
+commit;
+""".format(update_sql, update_sql), (appname2, 2, 1))
+
+    lock_count_sql = """SELECT COUNT(1)
+   FROM  pg_catalog.pg_locks         blocked_locks
+    JOIN pg_catalog.pg_stat_activity blocked_activity  ON blocked_activity.pid = blocked_locks.pid
+    JOIN pg_catalog.pg_locks         blocking_locks 
+        ON blocking_locks.locktype = blocked_locks.locktype
+        AND blocking_locks.DATABASE IS NOT DISTINCT FROM blocked_locks.DATABASE
+        AND blocking_locks.relation IS NOT DISTINCT FROM blocked_locks.relation
+        AND blocking_locks.page IS NOT DISTINCT FROM blocked_locks.page
+        AND blocking_locks.tuple IS NOT DISTINCT FROM blocked_locks.tuple
+        AND blocking_locks.virtualxid IS NOT DISTINCT FROM blocked_locks.virtualxid
+        AND blocking_locks.transactionid IS NOT DISTINCT FROM blocked_locks.transactionid
+        AND blocking_locks.classid IS NOT DISTINCT FROM blocked_locks.classid
+        AND blocking_locks.objid IS NOT DISTINCT FROM blocked_locks.objid
+        AND blocking_locks.objsubid IS NOT DISTINCT FROM blocked_locks.objsubid
+        AND blocking_locks.pid != blocked_locks.pid
+    JOIN pg_catalog.pg_stat_activity blocking_activity ON blocking_activity.pid = blocking_locks.pid
+   WHERE NOT blocked_locks.GRANTED 
+   	AND blocking_activity.application_name = %s 
+   	AND blocked_activity.application_name = %s """
+
+    is_locked = wait_on_result(
+        cursor = cursor, sql = lock_count_sql, binds = (appname1, appname2), 
+        expected_value = 1 )
+
+    if not is_locked:
+        raise Exception("ERROR: Couldn't reproduce a deadlock. That can happen on an extremely overloaded system.")
```

Comment:
Did this ever happen in CI while you were iterating on this PR? I'm just curious if we're risking adding a flaky test or if the cases where this would fail would be local environments where things are a bit less constrained/controlled.

Reply:
Yes - on an EC2 t2.micro instance where the CPUs credits where exhausted.

It's always possible to reproduce the deadlock, but pg_stat_database.deadlocks is updated asynchronously. The question is: for how long are we willing to wait on this update to happen. I hardcoded a generous limit of 10s, but under extreme conditions it might happen that pg_stat_database wasn't updated until this time. In such a case, the check would fail.

Which option do you prefer?
1. Keep it as it is.
2. Don't impose any limits.
3. Leave the check out.

Reply:
I think it can be kept as-is as long as the integrations-core CI itself isn't flaky.